Normalized Paths and how JSONPath relates to JSON Pointer (#152)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

JSONPath defines a string syntax for identifying values
within a JSON value.

JSONPath is not intended as a replacement for, but as a more powerful
companion to, JSON Pointer {{RFC6901}}. See {{json-pointer}}.

> 

Reviewer: Overall a great approach ... thanks.

  While this list can be represented in JSON, e.g. as an array, the
  nodelist is an abstract concept unrelated to JSON values.

> Normalized Path:

Reviewer: I definitely prefer 'Normalized Path'.

> "Normalization is the process of organizing a database to reduce redundancy and improve data integrity.
Normalization also simplifies the database design so that it achieves the optimal structure composed of atomic elements (i.e. elements that cannot be broken down into smaller parts)." 

https://database.guide/what-is-normalization/

>  In this context, a canonical form is a representation such that every object has a unique representation.

https://en.wikipedia.org/wiki/Canonical_form#Computing

So 'normalized path' in its canonical form is much more concrete than merely 'output path' . Its syntax definition is:

```
normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")

positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
```

Reviewer: A 'normalized path' usually is an 'output path'. It is frequently used as 'json-path' for revisiting previous selections (post processing ?) and is understood natively as accessor at least in JavaScript.

Author: Fair enough. I'm going to try using normalized path throughout. Please see the second commit.

Reviewer: hmm ... `quoted-member-name` must be replaced by `single-quoted-member-name` above.

Reviewer 2: > hmm ... `quoted-member-name` must be replaced by `single-quoted-member-name` above.

Indeed, but that is not everything we need to go fully canonical (https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/152#discussion_r796785776).

: A simple form of JSONPath expression that identifies a node by
  providing a query that results in exactly that node.  Similar
  to, but syntactically different from, a JSON Pointer {{-pointer}}.

>   Note: Normalized Paths are not canonical: two distinct Normalized Paths

Reviewer 2: This cries for an example.
It probably would be useful to have a place where normalized paths are discussed that is outside the definitions.

Author: Ok, I'll bite.

$['store']['book'][0]['title']
~~~~

to build paths that are input to a JSONPath implementation.
Bracket notation is more general than dot notation and can be used to construct
Normalized Paths (for instance, when a JSONPath implementation outputs Normalized Paths

> instead of the values identified by these paths).

Reviewer 2: not necessarily "instead of".

```suggestion
instead of or in addition to the values identified by these paths).
```

| `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
| `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |

## Normalized Paths

A Normalized Path identifies a node by providing a query that results in exactly that node. For example,
the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
`$.book[3]` and `$.book[5]`.

> 

Reviewer 2: `$['book'][3]` and `$["book"][3]` address the same node, so the Normalized Path is canonical in structure only, not in syntax.

> 

Reviewer: No ... even here we should state in the spec, that `$['book'][3]` is the one and only Normalized Path. Reason here is to reuse normalized pathes as strings in JSON documents without necessity of escaping ... exactly the same reason, why we support `'...'` additionally  to JSON's `"..."`.

> 

Reviewer 2: I like the idea of defining the preferred (canonical) syntax.  We still need to escape backslashes and decide how certain characters are shown (e.g., \r vs. \u000d, no \/ etc.) -- should not be too hard, but needs to be done.

> 

Author: I'm not sure of the goal here, or what the WG consensus is likely to be, especially since Normalized Paths don't currently feature much in the spec. Should we take this discussion to the mailing list?

> 

Reviewer: I don't know, what the goal of a mailing list discussion would be.

IIRC normalized pathes as an alternative result (to values) of queries never have been questionable. It is also the only way to verify, that identical queries lead to identical results ... wouldn't be possible with ambiguous node values only ... it's key for tests!

But if you want ... take this discussion to the mailing list?

> 

Author: > I think this is a tangent, but a useful one. Should there be a single canonical form for a Normalized Path?

Yes, that's the question. But not everyone who can usefully contribute will be tuned in to this PR.

> 
> (The discussion about "Normalized Path" vs. "Output Path" is a bit confused, like a discussion whether somebody is an "MIT Alumnus" or "CEO of acme corp." -- these are just orthogonal, one is what it is, one is where it is used.)

I understand, but I don't see the need for "Output Path" as a separate term. We can simply talk about the output from an implementation etc.

> 

Reviewer 2: > I understand, but I don't see the need for "Output Path" as a separate term. We can simply talk about the output from an implementation etc.

I don't see a need to *introduce* this term either, but we can say that the output of a query can be values and/or paths, and these are then of course output values and output paths without further definition or capitalization.

Let's open the Normalized Path determinism ("canonicalization") issue on the mailing list, though; we can finish the rest of this PR while this is being discussed.

> 

Reviewer: 'Output Path' is not needed. Result of a JSONPath query is a list of nodes formatted as canonical Normalized Pathes.

> 

Author: @cabo @goessner "canonical Normalized Path" is a mouthful. Why not simply "Canonical Path"?

> 

Reviewer 2: Normalized Path.

(Don't use the C word.)

> 

Author: For the record, what's wrong with "canonical"?

> 

Reviewer 2: Generally evokes enormous amounts of pain in people who have ever tried to do c14n.

Also, unnecessary here, as we already use the term "normalized".

An example application of Normalized Paths is a JSONPath implementation which outputs Normalized Paths
instead of, or in addition to, the values identified by these paths.

Since bracket notation is more general than dot notation, it is convenient in constructing
Normalized Paths. However, Normalized Paths are not required to use bracket notation when dot notation is appropriate.

Normalized Paths are not necessarily canonical since two distinct Normalized Paths may identify the same node.
For example, `$.a` and `$['a']` are equivalent Normalized Paths. With certain JSON documents, `$[-1]` and `$[4]` may

> be equivalent Normalized Paths.

Reviewer 3: I'm not sure I agree with `$[-1]` being considered normalized.  In my opinion, a normalized path should be one that distinctly identifies a node regardless of the input JSON.  `$[-1]` will identify different nodes depending on the input.

Author: > I'm not sure I agree with `$[-1]` being considered normalized. In my opinion, a normalized path should be one that distinctly identifies a node regardless of the input JSON. `$[-1]` will identify different nodes depending on the input.

I have some reservations too, but I was trying to make clear one consequence of _not_ insisting on a canonical form of Normalized Path. However, I think Normalized Paths only make sense in the context of a known/given JSON document. If the document varies, then the node identified by _any_ path can be different (e.g. just replace the identified node in one document with a distinct node).

Author: `$[-1]` is now not allowed as part of a Normalized Path.

> 

Reviewer: Here
```
$['a']
$[4]
```
qualifies for a normalized Path only.

Reviewer: Please see ...
```
normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")

positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
```

from above ...

Reviewer 2: s/positiv/unsigned/  (0 is allowed, but not positive)

Reviewer: ... I wanted to avoid `non-negative-element-index` ... now I like `unsigned` ...

Reviewer 4: > ... I wanted to avoid `non-negative-element-index` ... now I like `unsigned-element-index` ...

I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.

Author: > Please see ...
> 
> ```
> normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")
> 
> positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
> ```
> 
> from above ...

Unless the mailing list pushes back, I plan to adopt this prescriptive approach.

Reviewer 2: > > ... I wanted to avoid `non-negative-element-index` ... now I like `unsigned-element-index` ...
> 
> I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.

I get this discussion often.  Non-negative is pedantic.  The range of actual array indexes actually is unsigned integer, so why not just say that.  (We also allow indexing with negative integers, but that means something different.)

Reviewer 4: > > I think `non-negative-element-index` is better. The domain of indices is signed integers. That domain is restricted to non-negative integers for normalized paths, it's not that they've suddenly become unsigned.
> 
> I get this discussion often. Non-negative is pedantic. The range of actual array indexes actually is unsigned integer, so why not just say that. (We also allow indexing with negative integers, but that means something different.)

In this context, "pedantic" is just another name for "clear". It's unclear what it means to mix "unsigned" and "negative" integers on the same domain. It doesn't translate well into programming languages.

Author: > Please see ...
> 
> ```
> normalized-path = root-selector *("[" (quoted-member-name / positiv-element-index) "]")
> 
> positive-element-index = ( "0" / (DIGIT1 *DIGIT) )
> ```
> 
> from above ...

`quoted-member-name` allows two sort of variability: single or double quoted strings and escapes or unescaped characters. Should we leave these open or attempt to tie them down so there is one and only one representation of any Normalized Path?

Reviewer 2: I sensed the direction of the discussion that we wanted to have deterministic Normalized Path results, i.e. two implementations would generate the same output paths for the same JSON value and query.  That would need us to pick the kind of quoting and the details of the escaping.  Deterministic results generally help in testing.  We might still decide that this is too much work.

Reviewer: Please see my correction above ...

> hmm ... quoted-member-name must be replaced by single-quoted-member-name above.

so it should read somewhat like ...
```
normalized-path = root-selector *("[" *(%x27 *single-quoted %x27) / positiv-element-index) "]")

unsigned-element-index = ( "0" / (DIGIT1 *DIGIT) )
```

Reviewer: @cabo: I think, the benefit of clean testing justifies the effort here. However, I cannot really judge the effort involved in documenting escaping.

Author: Ok, I suggest we mandate single quotes (to align with JSON) and escapes (except for \/ and \uxxxx, instead of which / and the unicode character for U+xxxx must be used, respectively). Is that a reasonable starting position?

Reviewer: ... but, isn't escaping already clearly documented with `single-quoted` already in the current draft ... ?

Author: > ... but, isn't escaping already clearly documented with `single-quoted` already in the current draft ... ?

No, because it permits at least two representations of any given character. I think we should try to nail this down so there is a unique representation.

Reviewer 2: To be clear, JSON uses double quotes, so we are using single quotes to evade further JSON escaping, except where we need to escape.  We can't use the actual characters for \u0000 to \u001f, so we should use \b\t\n\f\r or the \u00xx for the other 11 cases.  We also need `\\` and `\'`, but we don't need `\/`.  I'm sure there will be people who would want to escape anything including and above \u00ff as well.  Oh, and we need to specify the case of the hexchars.

Author: Gotcha.

> I'm sure there will be people who would want to escape anything including and above \u00ff as well.

I guess escaping this range is safer, even if it can be criticised as ASCII-centric. It wouldn't surprise me if mandating unescaped characters for that range caused problems in some older programming languages.

Reviewer: I was thinking a while about that escape thing. There is a strong relation to string comparison, see ...

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/117#issuecomment-909356265

Consensus here obviously was *code point* i.e. *byte string* comparison.

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/117#issuecomment-1014932216

In this case here we need not to deal with arbitrary text content, but rather with member names. This should be less crtitical then. 

So, if restricting to code point equality here again, is all in all consistent.

I tried to collect some common scenarios:

1. Query results are reused on the same JSON source document for revisiting nodes with the purpose of ...
2. Normalized pathes are applied to similar JSON documents (output from a single application or into a standardised JSON format) by different users with different implementations on different platforms.
3. Compare query results from different implementations for test purposes.

Discussion:
1. Should be mostly applied in practise (estimated). Is uncritical regarding escapes, since JSON document, implementation, platform remains the same.
2. Is the most critical from my point of view. But the probability of dealing with escaped characters in this scenario is considered rather low.
3. For developers or advanced users only. Should be uncritical regarding escapes.

Are there other relevant scenarios, that I might be missing here?

So Glyn's approach or code point equality seems to be a pragmatic solution here.

| `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
| `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |

## Normalized Paths

A Normalized Path identifies a node by providing a query that results in exactly that node. For example,
the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths

> `$.book[3]` and `$.book[5]`.

Reviewer: ... `$['book'][3]` and  `$['book'][5]` ...

Author: Nice catch. I'll fix that.

An example application of Normalized Paths is a JSONPath implementation which outputs Normalized Paths
instead of, or in addition to, the values identified by these paths.

Since bracket notation is more general than dot notation, it is used to construct Normalized Paths.
the syntax of Normalized Paths is restricted so that there is one and only one way of representing any
given Normalized Path.

~~~~ abnf
normalized-path           = root-selector *(S normal-index-selector)
normal-index-selector     = "[" S (normal-quoted-member-name / normal-element-index) S "]"
normal-quoted-member-name = %x27 *normal-single-quoted %x27 ; 'string'
normal-single-quoted      = normal-unescaped /
                            ESC normal-escapable
normal-unescaped          = %x20-26 /                       ; omit control codes
                            %x28-5B /                       ; omit '
                            %x5D-7E                         ; omit \

>                             %xA0-FF                         ; omit control codes

Reviewer 2: We probably want to exclude U+00FF (Rubout).

>                             %xA0-FF                         ; omit control codes

Reviewer 2: I'm not sure how basic latin is better than other beyond-ASCII parts of Unicode.

Reviewer 2: I don't know how to explain that we need to escape Ё, ⽔, Č, ⨸, but not ÿ or ÷ or ¶.

Author: Is it simply a compression chosen because of the high frequency of latin characters in extant JSON?

Reviewer 2: But then you'd need to include all Latin characters.   I'm sure Mališa Vučinić would agree :-)
And what is better about Latin than Greek, Cyrillic, ...

Author: What would you propose? Escape almost all characters?

Reviewer 2: The reason I didn't propose anything yet is that I have no idea.  

Escaping *something* seems in line with https://nvd.nist.gov/vuln/detail/CVE-2021-42574 -- if we assume Normalized Paths are for humans to look at.

If this is all about programs, there is no point in escaping anything (except what is needed to stay similar enough to JSON).

normal-escapable           = (%x75 hexchar) ;  uXXXX U+XXXX ; TODO: exclude normal-unescaped codepoints
normal-element-index       = "0" / (DIGIT1 *DIGIT)          ; non-negative decimal integer

> ~~~~

Reviewer: ... I like that ...

Reviewer 2: So we'd have to use \u0027 for '?

Reviewer: Wouldn't we also need to omit `"` also, since we want to use normalized pathes - as they are - as potential text in JSON documents ?

Author: > So we'd have to use \u0027 for '?

Yes.

Author: > Wouldn't we also need to omit `"` also, since we want to use normalized pathes - as they are - as potential text in JSON documents ?

The requirement to use Normalized Paths as text in JSON documents is new to me. So far, I have only considered a JSONPath implementation outputting Normalized Paths instead of, or in addition to, values - end of story. What use case do you have in mind?

Reviewer: As I already commented here ...

https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/152/files#r796782971

charme of using `'` delimiters within JSONPath queries and resulting pathes is the possibility of having both as unescaped values in JSON documents. Use them for referencing locations in the same or other JSON documents in JSONPath context or simply document query results in testing contexts in JSON, whatever ... thus JSONPath and JSON harmonize perfectly together.

Author: Great. Let me build that in. Thanks!

Reviewer: example ... very unlikely ...

```json
{
     "j\"org" : "name of Jörg"
}
```

JSONPath result of `$['j"org']` or something similiar might / will be  `$['j"org']`. Putting them in 
```json
{
     "JSONPath_result": "$['j"org']"
}
```
will collide with JSON's  `"`.

Reviewer 2: Well, putting something into a JSON string will take care of that.

Reviewer 2: The charm of using `'` is that no escaping is needed, so the path stays readable in a JSON context.

Author: > As I already commented here ...
> 
> https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/152/files#r796782971
> 
> charme of using `'` delimiters within JSONPath queries and resulting pathes is the possibility of having both as unescaped values in JSON documents. Use them for referencing locations in the same or other JSON documents in JSONPath context or simply document query results in testing contexts in JSON, whatever ... thus JSONPath and JSON harmonize perfectly together.

@goessner Please see the latest commit which addresses your comment.

| `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
| `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |

## Normalized Paths

> A Normalized Path is a JSONPath that identifies a node by providing a query that results in exactly that node. For example,

Author: Hmmm. Usually the word "unique" (and its variants) introduces confusion into specs. The questions it raises are:

* What thing is unique?
* What is the set of things among which the thing is unique?

In this context, "uniquely identifies a node" could be taken a couple of ways:

1. A Normalized Path is the only JSONPath (with certain characteristics) that identifies a particular node. This is true.
2. A Normalized Path identifies a unique node. This is not true if we are allowed to vary the JSON document.

As for "eineindeutig", I understand this to refer to the mathematical concept of a one-one correspondence. For a given JSON document, there is indeed a one-one correspondence between its nodes and the Normalized Paths that identify these nodes. But the above proposal doesn't talk about a given JSON document, in which case the one-one correspondence doesn't make much sense.

I think we need an informal definition of Normalized Path so readers have a chance of engaging. We could follow it up with some more text along the lines of:

> For a given JSON document,  there is a one to one correspondence between the document's nodes and the Normalized Paths that identify these nodes.

> A Normalized Path is a JSONPath that identifies a node by providing a query that results in exactly that node. For example,

Reviewer: Yes ... this one to one thing is perfect ...

thanks again

> A Normalized Path is a JSONPath that identifies a node by providing a query that results in exactly that node. For example,

Author: Adopted. Thanks.

the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
`$['book'][3]` and `$['book'][5]`. For a given JSON document, there is a one to one correspondence between the document's
nodes and the Normalized Paths that identify these nodes.

A JSONPath implementation may output Normalized Paths instead of, or in addition to, the values identified by these paths.

Since bracket notation is more general than dot notation, it is used to construct Normalized Paths.
Single quotes are used to delimit string member names. This reduces the number of characters that
need escaping when Normalized Paths appear as strings (which are delimited with double quotes) in JSON documents.

The syntax of Normalized Paths is restricted so that there is one and only one way of representing any

> given Normalized Path. Putting this another way, for any two distinct Normalized Paths, there is a JSON document

Reviewer 2: ```suggestion
given Normalized Path. Putting this another way, for any two distinct Normalized Paths, a JSON document exists such that
it will yield distinct results when the Normalized Paths are applied to it.
```

Author: I pushed a commit with a variation of that suggestion.

| `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
| `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |

## Normalized Paths

A Normalized Path is a JSONPath with restricted syntax that identifies a node by providing a query that results in exactly that node. For example,
the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
`$['book'][3]` and `$['book'][5]`. For a given JSON document, there is a one to one correspondence between the document's
nodes and the Normalized Paths that identify these nodes.

An example application of Normalized Paths is a JSONPath implementation which outputs Normalized Paths

> instead of, or in addition to, the values identified by these paths.

Reviewer: A JSONPath implementation may output Normalized Paths
instead of, or in addition to, the values identified by these paths

Author: Adopted. Thanks.

Since bracket notation is more general than dot notation, it is used to construct Normalized Paths.
Single quotes are used to delimit string member names. This reduces the number of characters that
need escaping when Normalized Paths appear as strings (which are delimited with double quotes) in JSON documents.

The syntax of Normalized Paths is restricted so that there is one and only one way of representing any
given Normalized Path. Putting this another way, for any two distinct Normalized Paths, there is a JSON document

> which will yield distinct results when the Normalized Paths are applied to the document.

Reviewer: The syntax of Normalized Paths is restricted so that there is one and only one way of representing any
given node.

Reviewer: Putting this another way, any two distinct Normalized Paths will yield distinct results when applied to the any document.
...
Is this true? Both pathes may return '5' or 'undefined'.

Author: I deliberately phrased that sentence that way to avoid dealing with the case where a Normalized Path doesn't identify a node in a particular document.

May I suggest we merge the PR as-is and take up any further improvements separately as I think it represents the current consensus and we are mainly talking about editorial and descriptive tweaks.

Reviewer 2: (Ok, maybe s/is/exists/ makes this clearer as a mathematically phrased statement.)

Author: Yeah, that will clue in anyone familiar with predicate logic.

| `[3,4,5]` | `$[?index(@)==2]`<br>`$[?index(@)==17]` | `[5]`<br>`[]` | Select array element |
| `{"a":{"b":{5},c:0}}` | `$[?@.b==5 && !@.c]` | `[{"b":{5},c:0}]` | Existence  |

## Normalized Paths

A Normalized Path identifies a node by providing a query that results in exactly that node. For example,
the JSONPath expression `$.book[?(@.price<10)]` could select two values with Normalized Paths
`$['book'][3]` and `$['book'][5]`.

An example application of Normalized Paths is a JSONPath implementation which outputs Normalized Paths
instead of, or in addition to, the values identified by these paths.

Since bracket notation is more general than dot notation, it is used to construct Normalized Paths.

> Single quotes are used to delimit string member names. This reduces the number of characters which

Reviewer 2: ```suggestion
Single quotes are used to delimit string member names. This reduces the number of characters that
```

need escaping when Normalized Paths appear as strings (which are delimited with double quotes) in JSON documents.

The syntax of Normalized Paths is restricted so that there is one and only one way of representing any
given Normalized Path. Certain characters are escaped in one and only one way; all other
characters are unescaped.

~~~~ abnf

> normalized-path           = root-selector *(S normal-index-selector)

Reviewer 2: Do we actually want to allow spaces here?
That makes the output non-deterministic again.

Author: Nope. Good catch.

Reviewer: I do agree to **not** allow them.

normal-quoted-member-name = %x27 *normal-single-quoted %x27 ; 'string'
normal-single-quoted      = normal-unescaped /
                            ESC normal-escapable
normal-unescaped          = %x20-26 /                       ; omit control codes
                            %x28-5B /                       ; omit '
                            %x5D-10FFFF                     ; omit \
normal-escapable          = ( %x62 / %x66 / %x6E / %x72 / %x74 / ; \b \f \n \r \t
                                ; b /         ;  BS backspace U+0008
                                ; t /         ;  HT horizontal tab U+0009
                                ; n /         ;  LF line feed U+000A
                                ; f /         ;  FF form feed U+000C
                                ; r /         ;  CR carriage return U+000D
                                "'" /         ;  ' apostrophe U+0027
                                "\" /         ;  \ backslash (reverse solidus) U+005C
                                (%x75 normal-hexchar) ;  certain values u00XX U+00XX
                            )
normal-hexchar            = "0" "0"
                            (

>                               ("0" "0") /

Reviewer 2: This could be shortened as ("0" %x30-37)

Author: Agreed, but I think the longer form may be easier to read, so I'm going to punt on this for now.

                              ("0" "1") /
                              ("0" "2") /
                              ("0" "3") /
                              ("0" "4") /
                              ("0" "5") /
                              ("0" "6") /
                              ("0" "7") /

>                               ("0" "B") / ; omit U+0008-U+000A

Reviewer 2: These are case-insensitive.  We should decide on lower case or upper case.  Since we'll need to do a %x then, we can also %x65-66 again.

Author: Agreed. I'll plump for lower case. (Also, the use of `\b` etc. rather than `\B` is a mild precedent.)

## Normalized Paths

> A Normalized Path is a JSONPath which identifies a node by providing a query that results in exactly that node. For example,

Reviewer 2: ```suggestion
A Normalized Path is a JSONPath that identifies a node by providing a query that results in exactly that node. For example,
```

Author: Ok. Must get the hang of writing English some time.

the syntax of Normalized Paths is restricted so that there is one and only one way of representing any
given Normalized Path.

Single quotes are used to delimit string member names. This makes it possible to use Normalized Paths

> as strings in JSON documents (delimited with double quotes) without needing to escape the contents of

Reviewer 2: That is not true.  You still need to escape some of the contents of the strings when putting them into JSON; you just get `'` for free (and stay reasonably readable that way).

Reviewer 2: `"`, I think.

Reviewer: Yes of course that must be done manually ... but it is very unlikely to meet member names in any JSON document with embedded `"` (bad behavior).

Reviewer: what about pragmatically reusing the same ABNF, that rfc8259 uses for member names ... adding unescaped `'` of course ?

Reviewer 2: I think we already do...

```
      member = string name-separator value
```

where "string" is called "string-literal".
RFC 8259 hasn't done the work to avoid unpaired surrogates, so going back to their definition of "string" would be a regression.

But maybe I don't understand what you are trying to do.

Reviewer: oh ... I would like to mention the fact somewhere, that normalized pathes should be naturally the most performant queries in any implementation, compared with other - more complex - JSONPath queries leading to the same result node.

Reviewer: > But maybe I don't understand what you are trying to do.

you did ... thanks for the clarification.

Reviewer 5: You're correct that 8259 doesn't rule out naked surrogates and suchlike because the consensus was we couldn't retroactively ban things that were previously JSON per previous specifications. That's what I-JSON is for.  So the question is, should jsonpath member names conform to JSON or I-JSON?

Reviewer 2: I-JSON.

Reviewer 2: 8259 is schizophrenic: It requires UTF-8 for unescaped, but doesn't exclude unpaired surrogates for escaped.  That clearly would have been fixed if we had been a bit more audacious at the time.

Reviewer 5: OK, whether jsonpath should be more restrictive than JSON in what names it allows is an issue we're going to have to take up on its own.

Reviewer 2: As a WG chair, you may have to say something like this.

From a technical point of view, I find the suggestion that we could do anything that could be understood to lend additional credibility to the Zombie that Section 8.2 of RFC 8259 is, simply untenable. That isn't JSON, it is crimes of the past that wander the unlit cemeteries waiting to finally be buried.

the strings.

~~~~ abnf
normalized-path           = root-selector *(S normal-index-selector)
normal-index-selector     = "[" S (normal-quoted-member-name / normal-element-index) S "]"
normal-quoted-member-name = %x27 *normal-single-quoted %x27 ; 'string'
normal-single-quoted      = normal-unescaped /
                            ESC normal-escapable
normal-unescaped          = %x20-21 /                       ; omit control codes
                            %x23-26 /                       ; omit "
                            %x28-5B /                       ; omit '

>                             %x5D-7E                         ; omit \

Reviewer 2: I would comment on the missing `/`, but I think we'll need to get rid of the following line anyway (or change it considerably).

Author: What's interesting about `/`?

Reviewer 2: There is none in the current ABNF between %x5D-7E and %xA0-FF :-)

Author: Ah, gotcha. I was so focussed on the character ranges, I couldn't see the missing `/`. Fixed.

  addressed by the previous path fragment. Array indices always start
  at 0.

# JSON Pointer

This appendix is informative.

JSONPath is not intended as a replacement for, but as a more powerful
companion to, JSON Pointer {{RFC6901}}. The purposes of the two standards
are different.

JSON Pointer is for identifying a single value within a JSON document whose
structure is known.

JSONPath can identify a single value within a JSON document, for example by
using a Normalized Path. But JSONPath is also a query syntax that can be used
to search for and extract multiple values from JSON documents whose structure

> is partially known.

Reviewer 2: partially sounds like one half is (fully) known and the other half is completely unknown.

```suggestion
is known only in a general way.
```

> 

Reviewer 2: I think we should say something about conversion between JSON Pointer and JSONPath.

```
A Normalized JSONPath can be converted into a JSON Pointer by converting the syntax.
The inverse is not generally possible as a path component in a JSON Pointer that looks like a number may identify a member of a JSON object or may index an array.  For this conversion, knowledge of the structure of the references JSON value is needed.

  addressed by the previous path fragment. Array indices always start
  at 0.

# JSON Pointer

This appendix is informative.

JSONPath is not intended as a replacement for, but as a more powerful
companion to, JSON Pointer {{RFC6901}}. The purposes of the two standards
are different.

JSON Pointer is for identifying a single value within a JSON document whose
structure is known.

JSONPath can identify a single value within a JSON document, for example by
using a Normalized Path. But JSONPath is also a query syntax that can be used
to search for and extract multiple values from JSON documents whose structure
is known only in a general way.

A Normalized JSONPath can be converted into a JSON Pointer by converting the syntax,
without knowledge of any JSON document. The inverse is not generally true: a numeric
path component in a JSON Pointer may identify a member of a JSON object or may index an array.

> For this conversion, knowledge of the structure of the JSON document is needed.

Reviewer 2: ```suggestion
For conversion to a JSONPath query, knowledge of the structure of the JSON document is needed to distinguish these cases.
```

Author: I'll include this in the next commit, thanks.
